### PR TITLE
stop casting hidden states to float32 for cut cross entropy

### PR DESCRIPTION
**Pull request by mmathew23**
T4 currently has an issue training lm_head's since bfloat16 is not supported. A few users have experienced the same issue and it's reproducible. 

[#2171](https://github.com/unslothai/unsloth/issues/2127)
[#2253](https://github.com/unslothai/unsloth/issues/2253)

[Original PR](https://github.com/unslothai/unsloth/pull/1200) casts lm_head to float32 since it's a module to save when training the lm head with lora. Before going to cut cross entropy, the output embeddings are casted to the same dtype as lm_head. Cut cross entropy specifically asserts that the embeddings are float16 or bfloat16 and raises an AssertionError if something else like float32. It's not clear why embeddings can't be float32, but I presume it's due to code not being optimized for that type.

This change only allows casting hidden_states to the weight's dtype if it's one of float16 or bfloat16. Otherwise it keeps the original dtype. Even though this code touches almost all unsloth finetuning usecases it's a low risk change since cut cross entropy would fail anyways if the dtype wasn't one of float16 or bfloat16.

### CODE REVIEW

**Comment by Datta0** [comment]
Hey @mmathew23 can you please verify if the losses match up to HF's implementation.
iirc, the initial dtypes were set that way to match the losses....

**Comment by mmathew23** [comment]
Hi @Datta0 here's a comparison script and the results.

```
import unsloth
from unsloth_zoo.loss_utils import fused_linear_cross_entropy
from typing import Optional
from itertools import product
import gc
import torch
from transformers.loss.loss_utils import ForCausalLMLoss

# %%

def cleanup(*tensors):
    for tensor in tensors:
        if tensor.grad is not None:
            tensor.grad.zero_()
    for _ in range(3):
        gc.collect()
        torch.cuda.empty_cache()

def get_promoted_dtype(tensor1, tensor2):
    # Promote dtype manually
    dtypes = {tensor1.dtype, tensor2.dtype}
    if torch.float32 in dtypes:
        return torch.float32
    elif torch.bfloat16 in dtypes:
        return torch.bfloat16
    else:
        return torch.float16

def hf_linear_cross_entropy(
    hidden_states      : torch.Tensor,
    lm_weight          : torch.Tensor,
    labels             : torch.Tensor,
    num_items_in_batch : int = None,
    ignore_index       : int = -100,
    reduction          : str = "mean",
    logit_softcapping  : float = 0,
    accuracy_threshold : str = "auto",
    shift_labels: Optional[torch.Tensor] = None,
):
    with torch.amp.autocast(device_type="cuda", dtype=get_promoted_dtype(hidden_states, lm_weight)):
        logits = torch.matmul(hidden_states, lm_weight.T)
    vocab_size = lm_weight.shape[0]
    loss = ForCausalLMLoss(logits, labels, vocab_size, num_items_in_batch=num_items_in_batch, ignore_index=ignore_index, reduction=reduction, shift_labels=shift_labels)
    return loss

# %%

torch.manual_seed(3407)

# %%

vocab_size = 128256//2
hidden_dim = 2048
seq_len = 1024
device="cuda"
embeddingsfp32 = torch.randn(2, seq_len, hidden_dim, dtype=torch.float32, device=device)
embeddingsfp16 = embeddingsfp32.clone().to(torch.float16)
embeddingsbf16 = embeddingsfp32.clone().to(torch.bfloat16)
embeddings = [embeddingsfp32, embeddingsfp16, embeddingsbf16]

labels = torch.randint(0, vocab_size, (2, seq_len), device=device)

lm_weightfp32 = torch.randn(vocab_size, hidden_dim, dtype=torch.float32, device=device)
lm_weightfp16 = lm_weightfp32.clone().to(torch.float16)
lm_weightbf16 = lm_weightfp32.clone().to(torch.bfloat16)
lm_weights = [lm_weightfp32, lm_weightfp16, lm_weightbf16]

reductions = ["mean"]

# %%

results = {}
for embedding, lm_weight, reduction in product(embeddings, lm_weights, reductions):
    key = f"{embedding.dtype}_{lm_weight.dtype}"
    hf_embedding = embedding.clone().detach().requires_grad_(True)
    hf_lm_weight = lm_weight.clone().detach().requires_grad_(True)
    results[key] = {}
    try:
        loss = hf_linear_cross_entropy(hf_embedding, hf_lm_weight, labels, reduction=reduction)
        results[key]["hf_loss"] = loss.item()
    except Exception as e:
        print(f"HF failed for {key}: {e}")
        results[key]["hf_loss"] = 0
    try:
        loss.backward()
        results[key]["hf_emb_grad_norm"] = hf_embedding.grad.norm().item()
        results[key]["hf_lm_weight_grad_norm"] = hf_lm_weight.grad.norm().item()
    except Exception as e:
        print(f"HF backward failed for {key}: {e}")
        results[key]["hf_emb_grad_norm"] = 0
        results[key]["hf_lm_weight_grad_norm"] = 0

    try:
        del loss
    except Exception as e:
        print(f"HF cleanup failed for {key}: {e}")

    cleanup(hf_embedding, hf_lm_weight)

    unsloth_embedding = embedding.clone().detach().requires_grad_(True)
    unsloth_lm_weight = lm_weight.clone().detach().requires_grad_(True)
    try:
        loss = fused_linear_cross_entropy(unsloth_embedding, unsloth_lm_weight, labels, reduction=reduction)
        results[key]["unsloth_loss"] = loss.item()
    except Exception as e:
        print(f"Unsloth failed for {key}: {e}")
        results[key]["unsloth_loss"] = 0

    try:
        loss.backward()
        results[key]["unsloth_emb_grad_norm"] = unsloth_embedding.grad.norm().item()
        results[key]["unsloth_lm_weight_grad_norm"] = unsloth_lm_weight.grad.norm().item()
    except Exception as e:
        print(f"Unsloth backward failed for {key}: {e}")
        results[key]["unsloth_emb_grad_norm"] = 0
        results[key]["unsloth_lm_weight_grad_norm"] = 0
    try:
        del loss
    except Exception as e:
        print(f"Unsloth cleanup failed for {key}: {e}")

    cleanup(unsloth_embedding, unsloth_lm_weight)


# %%

from cut_cross_entropy import linear_cross_entropy

def fused_linear_cross_entropy_fixed(
    hidden_states      : torch.Tensor,
    lm_weight          : torch.Tensor,
    labels             : torch.Tensor,
    num_items_in_batch : int = None,
    ignore_index       : int = -100,
    reduction          : str = "mean",
    logit_softcapping  : float = 0,
    accuracy_threshold : str = "auto",
):
    # All Unsloth Zoo code licensed under LGPLv3
    reduction = "sum" if num_items_in_batch is not None else "mean"
    if logit_softcapping == 0: logit_softcapping = None
    loss = linear_cross_entropy(
        hidden_states.to(lm_weight.dtype if lm_weight.dtype in (torch.float16, torch.bfloat16) else hidden_states.dtype),
        lm_weight,
        targets      = labels,
        ignore_index = ignore_index,
        softcap      = logit_softcapping,
        reduction    = reduction,
        shift        = True,
        filter_eps   = accuracy_threshold,
    )
    if num_items_in_batch is not None: loss = loss / num_items_in_batch
    return loss
pass

# %%

results_post_fix = {}
for embedding, lm_weight, reduction in product(embeddings, lm_weights, reductions):
    key = f"{embedding.dtype}_{lm_weight.dtype}"
    hf_embedding = embedding.clone().detach().requires_grad_(True)
    hf_lm_weight = lm_weight.clone().detach().requires_grad_(True)
    results_post_fix[key] = {}
    try:
        loss = hf_linear_cross_entropy(hf_embedding, hf_lm_weight, labels, reduction=reduction)
        results_post_fix[key]["hf_loss"] = loss.item()
    except Exception as e:
        print(f"HF failed for {key}: {e}")
        results_post_fix[key]["hf_loss"] = 0
    try:
        loss.backward()
        results_post_fix[key]["hf_emb_grad_norm"] = hf_embedding.grad.norm().item()
        results_post_fix[key]["hf_lm_weight_grad_norm"] = hf_lm_weight.grad.norm().item()
    except Exception as e:
        print(f"HF backward failed for {key}: {e}")
        results_post_fix[key]["hf_emb_grad_norm"] = 0
        results_post_fix[key]["hf_lm_weight_grad_norm"] = 0

    try:
        del loss
    except Exception as e:
        print(f"HF cleanup failed for {key}: {e}")

    cleanup(hf_embedding, hf_lm_weight)

    unsloth_embedding = embedding.clone().detach().requires_grad_(True)
    unsloth_lm_weight = lm_weight.clone().detach().requires_grad_(True)
    try:
        loss = fused_linear_cross_entropy_fixed(unsloth_embedding, unsloth_lm_weight, labels, reduction=reduction)
        results_post_fix[key]["unsloth_loss"] = loss.item()
    except Exception as e:
        print(f"Unsloth failed for {key}: {e}")
        results_post_fix[key]["unsloth_loss"] = 0

    try:
        loss.backward()
        results_post_fix[key]["unsloth_emb_grad_norm"] = unsloth_embedding.grad.norm().item()
        results_post_fix[key]["unsloth_lm_weight_grad_norm"] = unsloth_lm_weight.grad.norm().item()
    except Exception as e:
        print(f"Unsloth backward failed for {key}: {e}")
        results_post_fix[key]["unsloth_emb_grad_norm"] = 0
        results_post_fix[key]["unsloth_lm_weight_grad_norm"] = 0
    try:
        del loss
    except Exception as e:
        print(f"Unsloth cleanup failed for {key}: {e}")

    cleanup(unsloth_embedding, unsloth_lm_weight)


# %%

from typing import Dict
from tabulate import tabulate

def compare_structured(dict1: Dict, dict2: Dict):
    metrics = [
        ("loss", "hf_loss", "unsloth_loss"),
        ("emb_grad_norm", "hf_emb_grad_norm", "unsloth_emb_grad_norm"),
        ("lm_weight_grad_norm", "hf_lm_weight_grad_norm", "unsloth_lm_weight_grad_norm")
    ]

    for dtype_combo in dict1:
        print(f"### Dtype Combination: `{dtype_combo}`\n")

        table = []
        headers = [
            "Metric",
            "hf Dict1", "hf Dict2", "hf Δ",
            "unsloth Dict1", "unsloth Dict2", "unsloth Δ",
            "Δ (unsloth-hf D1)", "Δ (unsloth-hf D2)"
        ]

        for label, hf_key, unsloth_key in metrics:
            hf1 = dict1[dtype_combo].get(hf_key, 0)
            hf2 = dict2[dtype_combo].get(hf_key, 0)
            u1 = dict1[dtype_combo].get(unsloth_key, 0)
            u2 = dict2[dtype_combo].get(unsloth_key, 0)

            hf_diff = hf2 - hf1
            unsloth_diff = u2 - u1
            delta1 = u1 - hf1
            delta2 = u2 - hf2

            def fmt(val): return f"{val:.5f}"
            def fmt_diff(val): return f"**{val:+.5f}**" if abs(val) > 1e-5 else f"{val:+.5f}"

            table.append([
                label,
                fmt(hf1), fmt(hf2), fmt_diff(hf_diff),
                fmt(u1), fmt(u2), fmt_diff(unsloth_diff),
                fmt_diff(delta1), fmt_diff(delta2),
            ])

        print(tabulate(table, headers=headers, tablefmt="github"))
        print()


compare_structured(results, results_post_fix)
```

Ouput:
### Dtype Combination: `torch.float32_torch.float32`

| Metric              |   hf Dict1 |   hf Dict2 |   hf Δ |   unsloth Dict1 |   unsloth Dict2 |   unsloth Δ | Δ (unsloth-hf D1)   | Δ (unsloth-hf D2)   |
|---------------------|------------|------------|--------|-----------------|-----------------|-------------|---------------------|---------------------|
| loss                |  193.914   |  193.914   |      0 |         193.914 |         193.914 |           0 | **-0.00003**        | **-0.00003**        |
| emb_grad_norm       |    1.38433 |    1.38433 |      0 |           0     |           0     |           0 | **-1.38433**        | **-1.38433**        |
| lm_weight_grad_norm |    1.38082 |    1.38082 |      0 |           0     |           0     |           0 | **-1.38082**        | **-1.38082**        |

### Dtype Combination: `torch.float32_torch.float16`

| Metric              |   hf Dict1 |   hf Dict2 |   hf Δ |   unsloth Dict1 |   unsloth Dict2 |   unsloth Δ | Δ (unsloth-hf D1)   | Δ (unsloth-hf D2)   |
|---------------------|------------|------------|--------|-----------------|-----------------|-------------|---------------------|---------------------|
| loss                |  193.914   |  193.914   |      0 |       193.914   |       193.914   |           0 | **-0.00029**        | **-0.00029**        |
| emb_grad_norm       |    1.38433 |    1.38433 |      0 |         1.3843  |         1.3843  |           0 | **-0.00004**        | **-0.00004**        |
| lm_weight_grad_norm |    1.38086 |    1.38086 |      0 |         1.38086 |         1.38086 |           0 | +0.00000            | +0.00000            |

### Dtype Combination: `torch.float32_torch.bfloat16`

| Metric              |   hf Dict1 |   hf Dict2 |   hf Δ |   unsloth Dict1 |   unsloth Dict2 |   unsloth Δ | Δ (unsloth-hf D1)   | Δ (unsloth-hf D2)   |
|---------------------|------------|------------|--------|-----------------|-----------------|-------------|---------------------|---------------------|
| loss                |  193.914   |  193.914   |      0 |       193.915   |       193.915   |           0 | **+0.00026**        | **+0.00026**        |
| emb_grad_norm       |    1.38429 |    1.38429 |      0 |         1.38325 |         1.38325 |           0 | **-0.00104**        | **-0.00104**        |
| lm_weight_grad_norm |    1.38281 |    1.38281 |      0 |         1.38281 |         1.38281 |           0 | +0.00000            | +0.00000            |

### Dtype Combination: `torch.float16_torch.float32`

| Metric              |   hf Dict1 |   hf Dict2 |   hf Δ |   unsloth Dict1 |   unsloth Dict2 | unsloth Δ    | Δ (unsloth-hf D1)   | Δ (unsloth-hf D2)   |
|---------------------|------------|------------|--------|-----------------|-----------------|--------------|---------------------|---------------------|
| loss                |  193.914   |  193.914   |      0 |         193.914 |       193.913   | **-0.00082** | **-0.00002**        | **-0.00084**        |
| emb_grad_norm       |    1.38477 |    1.38477 |      0 |           0     |         1.38477 | **+1.38477** | **-1.38477**        | +0.00000            |
| lm_weight_grad_norm |    1.38082 |    1.38082 |      0 |           0     |         1.38079 | **+1.38079** | **-1.38082**        | **-0.00003**        |

### Dtype Combination: `torch.float16_torch.float16`

| Metric              |   hf Dict1 |   hf Dict2 |   hf Δ |   unsloth Dict1 |   unsloth Dict2 |   unsloth Δ | Δ (unsloth-hf D1)   | Δ (unsloth-hf D2)   |
|---------------------|------------|------------|--------|-----------------|-----------------|-------------|---------------------|---------------------|
| loss                |  193.913   |  193.913   |      0 |       193.914   |       193.914   |           0 | **+0.00043**        | **+0.00043**        |
| emb_grad_norm       |    1.38477 |    1.38477 |      0 |         1.38477 |         1.38477 |           0 | +0.00000            | +0.00000            |
| lm_weight_grad_norm |    1.38086 |    1.38086 |      0 |         1.38086 |         1.38086 |           0 | +0.00000            | +0.00000            |

### Dtype Combination: `torch.float16_torch.bfloat16`

| Metric              |   hf Dict1 |   hf Dict2 |   hf Δ |   unsloth Dict1 |   unsloth Dict2 |   unsloth Δ | Δ (unsloth-hf D1)   | Δ (unsloth-hf D2)   |
|---------------------|------------|------------|--------|-----------------|-----------------|-------------|---------------------|---------------------|
| loss                |  193.929   |  193.929   |      0 |       193.915   |       193.915   |           0 | **-0.01407**        | **-0.01407**        |
| emb_grad_norm       |    1.38379 |    1.38379 |      0 |         1.38281 |         1.38281 |           0 | **-0.00098**        | **-0.00098**        |
| lm_weight_grad_norm |    1.38281 |    1.38281 |      0 |         1.38281 |         1.38281 |           0 | +0.00000            | +0.00000            |

### Dtype Combination: `torch.bfloat16_torch.float32`

| Metric              |   hf Dict1 |   hf Dict2 |   hf Δ |   unsloth Dict1 |   unsloth Dict2 | unsloth Δ    | Δ (unsloth-hf D1)   | Δ (unsloth-hf D2)   |
|---------------------|------------|------------|--------|-----------------|-----------------|--------------|---------------------|---------------------|
| loss                |  193.913   |  193.913   |      0 |         193.913 |       193.913   | **-0.00024** | +0.00000            | **-0.00024**        |
| emb_grad_norm       |    1.38281 |    1.38281 |      0 |           0     |         1.38281 | **+1.38281** | **-1.38281**        | +0.00000            |
| lm_weight_grad_norm |    1.38086 |    1.38086 |      0 |           0     |         1.37975 | **+1.37975** | **-1.38086**        | **-0.00110**        |

### Dtype Combination: `torch.bfloat16_torch.float16`

| Metric              |   hf Dict1 |   hf Dict2 |   hf Δ |   unsloth Dict1 |   unsloth Dict2 |   unsloth Δ | Δ (unsloth-hf D1)   | Δ (unsloth-hf D2)   |
|---------------------|------------|------------|--------|-----------------|-----------------|-------------|---------------------|---------------------|
| loss                |  193.922   |  193.922   |      0 |       193.912   |       193.912   |           0 | **-0.01007**        | **-0.01007**        |
| emb_grad_norm       |    1.38281 |    1.38281 |      0 |         1.38281 |         1.38281 |           0 | +0.00000            | +0.00000            |
| lm_weight_grad_norm |    1.37988 |    1.37988 |      0 |         1.38086 |         1.38086 |           0 | **+0.00098**        | **+0.00098**        |


Most relevant are the last 3 columns. the last two columns are the delta's between hf and unsloth with random inputs and weights. As you can see there are small diffs even before the change. So if post fix the diffs remain the same there shouldn't be any impact. The only two scenarios where there is a diff after fix is when output embeddings are {float16, bfloat16} and lm_head weights are float32. In these two scenarios before the fix unsloth fails on the backward step but successful on the forward. The diff on the loss is small and makes sense. Prior we upcasted hidden states to float32 but not we are keeping in 16 bit so make sense we are off around the precision limit but it's very small. I think this diff is ok because backward completely fails, and now it works. 

The other notable scenario is when they are both float32. Unsloth will always fail on the backward because cut cross entropy requires the output embedding to be in float16 or bfloat16. Otherwise the results before and after remain consistent. The diffs between unsloth and hf I think are due to the fact unsloth compiles the loss so op order and specific gpu arch would dictate this.

Let me know if there are any questions or concerns.

**Comment by bbznop** [comment]
@mmathew23 Hello, why is this Pull Request closed? Is there something wrong?
